### PR TITLE
Generalize `Base._cat` to non-`Val`, typed `Base._cat_t` and implement `typed_hcat`, `typed_vcat`, `typed_hvcat`, `typed_hvncat`

### LINKED ISSUE #148: Missing `hcat`/`vcat`/`hvcat` dispatches
Hey, there! (cc @mofeing).  I have been using `Reactant` with `Tenet` and `Quac` to define a loss function like this:
```julia
function loss(params)
    tn = Tenet.TensorNetwork(expectation(params))

    tn′ = adapt(Reactant.ConcreteRArray, tn)
    path = einexpr(tn′; optimizer=Greedy())

    return contract(tn′)
end

function ∇loss(params)
    foo = Enzyme.gradient(ReverseWithPrimal, loss, params)
    return foo.val, foo.derivs
end
```

The `loss` function seems to work as expected, but if I try finding the gradient, I get some problems:
```julia
julia> ∇f = @compile ∇loss(params′)
┌ Warning: Performing scalar indexing on task Task (runnable) @0x00007cb4fb200010.
│ Invocation resulted in scalar indexing of a TracedRArray.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on device, but very slowly on the CPU,
│ and require expensive copies and synchronization each time and therefore should be avoided.
└ @ Reactant ~/.julia/packages/Reactant/FwbOv/src/TracedRArray.jl:53
ERROR: MethodError: Cannot `convert` an object of type Reactant.TracedRArray{Float64, 0} to an object of type Float64

Closest candidates are:
  convert(::Type{T}, ::Number) where T<:Number
   @ Base number.jl:7
  convert(::Type{T}, ::ConcreteRArray{T, 0}) where T
   @ Reactant ~/.julia/packages/Reactant/FwbOv/src/ConcreteRArray.jl:73
  convert(::Type{T}, ::T) where T
   @ Base Base.jl:84
  ...

Stacktrace:
  [1] Ry(θ::Reactant.TracedRArray{Float64, 0})
    @ Quac ~/.julia/packages/Quac/9bB1B/src/Gate.jl:137
  [2] Ry(; θ::Reactant.TracedRArray{Float64, 0})
    @ Quac ~/.julia/packages/Quac/9bB1B/src/Gate.jl:136
  [3] Ry(lanes::Int64; params::@Kwargs{θ::Reactant.TracedRArray{Float64, 0}})
    @ Quac ~/.julia/packages/Quac/9bB1B/src/Gate.jl:140
  [4] ansatz(params::Reactant.TracedRArray{Float64, 1})
    @ Main ./REPL[8]:3
  [5] expectation(params::Reactant.TracedRArray{Float64, 1})
    @ Main ./REPL[9]:6
  [6] loss
    @ ./REPL[12]:2 [inlined]
  [7] (::Tuple{})(none::Reactant.TracedRArray{Float64, 1})
    @ Base.Experimental ./<missing>:0
  [8] (::Reactant.var"#26#35"{typeof(loss), Reactant.MLIR.IR.Block, Vector{Reactant.TracedRArray}, Tuple{Reactant.TracedRArray{Float64, 1}}})()
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/utils.jl:100
  [9] block!(f::Reactant.var"#26#35"{typeof(loss), Reactant.MLIR.IR.Block, Vector{Reactant.TracedRArray}, Tuple{Reactant.TracedRArray{Float64, 1}}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Block.jl:201
 [10] make_mlir_fn(f::Function, args::Tuple{Reactant.TracedRArray{Float64, 1}}, kwargs::Tuple{}, name::String, concretein::Bool; toscalar::Bool)
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/utils.jl:74
 [11] make_mlir_fn
    @ ~/.julia/packages/Reactant/FwbOv/src/utils.jl:24 [inlined]
 [12] autodiff(::ReverseMode{true, false, FFIABI, false, true}, f::Const{typeof(loss)}, ::Type{Active}, args::Duplicated{Reactant.TracedRArray{Float64, 1}})
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/Interpreter.jl:253
 [13] autodiff
    @ ~/.julia/packages/Enzyme/8GSlk/src/Enzyme.jl:512 [inlined]
 [14] macro expansion
    @ ~/.julia/packages/Enzyme/8GSlk/src/Enzyme.jl:1705 [inlined]
 [15] gradient
    @ ~/.julia/packages/Enzyme/8GSlk/src/Enzyme.jl:1646 [inlined]
 [16] ∇loss
    @ ./REPL[13]:2 [inlined]
 [17] (::Tuple{})(none::Reactant.TracedRArray{Float64, 1})
    @ Base.Experimental ./<missing>:0
 [18] (::Reactant.var"#26#35"{typeof(∇loss), Reactant.MLIR.IR.Block, Vector{Reactant.TracedRArray}, Tuple{Reactant.TracedRArray{Float64, 1}}})()
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/utils.jl:100
 [19] block!(f::Reactant.var"#26#35"{typeof(∇loss), Reactant.MLIR.IR.Block, Vector{Reactant.TracedRArray}, Tuple{Reactant.TracedRArray{Float64, 1}}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Block.jl:201
 [20] make_mlir_fn(f::Function, args::Tuple{ConcreteRArray{Float64, 1}}, kwargs::Tuple{}, name::String, concretein::Bool; toscalar::Bool)
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/utils.jl:74
 [21] make_mlir_fn
    @ ~/.julia/packages/Reactant/FwbOv/src/utils.jl:24 [inlined]
 [22] #6
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:253 [inlined]
 [23] block!(f::Reactant.Compiler.var"#6#11"{typeof(∇loss), Tuple{ConcreteRArray{Float64, 1}}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Block.jl:201
 [24] #5
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:252 [inlined]
 [25] mmodule!(f::Reactant.Compiler.var"#5#10"{Reactant.MLIR.IR.Module, typeof(∇loss), Tuple{ConcreteRArray{Float64, 1}}}, blk::Reactant.MLIR.IR.Module)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Module.jl:93
 [26] compile_mlir!(mod::Reactant.MLIR.IR.Module, f::Function, args::Tuple{ConcreteRArray{Float64, 1}}; optimize::Bool)
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:249
 [27] compile_mlir!
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:248 [inlined]
 [28] (::Reactant.Compiler.var"#30#32"{typeof(∇loss), Tuple{ConcreteRArray{Float64, 1}}})()
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:576
 [29] context!(f::Reactant.Compiler.var"#30#32"{typeof(∇loss), Tuple{ConcreteRArray{Float64, 1}}}, ctx::Reactant.MLIR.IR.Context)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Context.jl:71
 [30] compile_xla(f::Function, args::Tuple{ConcreteRArray{Float64, 1}}; client::Nothing)
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:573
 [31] compile_xla
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:567 [inlined]
 [32] compile(f::Function, args::Tuple{ConcreteRArray{Float64, 1}}; client::Nothing)
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:600
 [33] compile(f::Function, args::Tuple{ConcreteRArray{Float64, 1}})
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:599
 [34] top-level scope
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:360
```

I tried with `Yao` instead of `Quac` and, once I fixed one first bug with the line:
```julia
Reactant.TracedRArray{Float64, 0}(t::Reactant.TracedRArray{Float64, 0}) = t
```

I got a similar error:
```julia
julia> loss(params)
0-dimensional Tensor{Float64, 0, Array{Float64, 0}}:
0.0

julia> ∇f = @compile ∇loss(params′)

┌ Warning: Performing scalar indexing on task Task (runnable) @0x0000744739600010.
│ Invocation resulted in scalar indexing of a TracedRArray.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on device, but very slowly on the CPU,
│ and require expensive copies and synchronization each time and therefore should be avoided.
└ @ Reactant ~/.julia/packages/Reactant/FwbOv/src/TracedRArray.jl:53
ERROR: MethodError: no method matching Reactant.TracedRArray{Float64, 0}(::Reactant.TracedRArray{Float64, 0})

Closest candidates are:
  Reactant.TracedRArray{T, N}(::Tuple, ::Union{Nothing, Reactant.MLIR.IR.Value}, ::Any) where {T, N}
   @ Reactant ~/.julia/packages/Reactant/FwbOv/src/TracedRArray.jl:9

Stacktrace:
  [1] RotationGate{2, Reactant.TracedRArray{Float64, 0}, YGate}(block::YGate, theta::Reactant.TracedRArray{Float64, 0})
    @ YaoBlocks ~/.julia/packages/YaoBlocks/DuDXk/src/primitive/rotation_gate.jl:22
  [2] RotationGate(block::YGate, theta::Reactant.TracedRArray{Float64, 0})
    @ YaoBlocks ~/.julia/packages/YaoBlocks/DuDXk/src/primitive/rotation_gate.jl:26
  [3] Ry(theta::Reactant.TracedRArray{Float64, 0})
    @ YaoBlocks ~/.julia/packages/YaoBlocks/DuDXk/src/primitive/rotation_gate.jl:59
  [4] ansatz(params::Reactant.TracedRArray{Float64, 1})
    @ Main ./REPL[10]:2
  [5] expectation(params::Reactant.TracedRArray{Float64, 1})
    @ Main ./REPL[11]:7
  [6] loss
    @ ./REPL[12]:2 [inlined]
  [7] (::Tuple{})(none::Reactant.TracedRArray{Float64, 1})
    @ Base.Experimental ./<missing>:0
  [8] (::Reactant.var"#26#35"{typeof(loss), Reactant.MLIR.IR.Block, Vector{Reactant.TracedRArray}, Tuple{Reactant.TracedRArray{Float64, 1}}})()
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/utils.jl:100
  [9] block!(f::Reactant.var"#26#35"{typeof(loss), Reactant.MLIR.IR.Block, Vector{Reactant.TracedRArray}, Tuple{Reactant.TracedRArray{Float64, 1}}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Block.jl:201
 [10] make_mlir_fn(f::Function, args::Tuple{Reactant.TracedRArray{Float64, 1}}, kwargs::Tuple{}, name::String, concretein::Bool; toscalar::Bool)
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/utils.jl:74
 [11] make_mlir_fn
    @ ~/.julia/packages/Reactant/FwbOv/src/utils.jl:24 [inlined]
 [12] autodiff(::ReverseMode{true, false, FFIABI, false, true}, f::EnzymeCore.Const{typeof(loss)}, ::Type{Active}, args::Duplicated{Reactant.TracedRArray{Float64, 1}})
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/Interpreter.jl:253
 [13] autodiff
    @ ~/.julia/packages/Enzyme/8GSlk/src/Enzyme.jl:512 [inlined]
 [14] macro expansion
    @ ~/.julia/packages/Enzyme/8GSlk/src/Enzyme.jl:1705 [inlined]
 [15] gradient
    @ ~/.julia/packages/Enzyme/8GSlk/src/Enzyme.jl:1646 [inlined]
 [16] ∇loss
    @ ./REPL[13]:2 [inlined]
 [17] (::Tuple{})(none::Reactant.TracedRArray{Float64, 1})
    @ Base.Experimental ./<missing>:0
 [18] (::Reactant.var"#26#35"{typeof(∇loss), Reactant.MLIR.IR.Block, Vector{Reactant.TracedRArray}, Tuple{Reactant.TracedRArray{Float64, 1}}})()
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/utils.jl:100
 [19] block!(f::Reactant.var"#26#35"{typeof(∇loss), Reactant.MLIR.IR.Block, Vector{Reactant.TracedRArray}, Tuple{Reactant.TracedRArray{Float64, 1}}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Block.jl:201
 [20] make_mlir_fn(f::Function, args::Tuple{ConcreteRArray{Float64, 1}}, kwargs::Tuple{}, name::String, concretein::Bool; toscalar::Bool)
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/utils.jl:74
 [21] make_mlir_fn
    @ ~/.julia/packages/Reactant/FwbOv/src/utils.jl:24 [inlined]
 [22] #6
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:253 [inlined]
 [23] block!(f::Reactant.Compiler.var"#6#11"{typeof(∇loss), Tuple{ConcreteRArray{Float64, 1}}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Block.jl:201
 [24] #5
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:252 [inlined]
 [25] mmodule!(f::Reactant.Compiler.var"#5#10"{Reactant.MLIR.IR.Module, typeof(∇loss), Tuple{ConcreteRArray{Float64, 1}}}, blk::Reactant.MLIR.IR.Module)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Module.jl:93
 [26] compile_mlir!(mod::Reactant.MLIR.IR.Module, f::Function, args::Tuple{ConcreteRArray{Float64, 1}}; optimize::Bool)
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:249
 [27] compile_mlir!
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:248 [inlined]
 [28] (::Reactant.Compiler.var"#30#32"{typeof(∇loss), Tuple{ConcreteRArray{Float64, 1}}})()
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:576
 [29] context!(f::Reactant.Compiler.var"#30#32"{typeof(∇loss), Tuple{ConcreteRArray{Float64, 1}}}, ctx::Reactant.MLIR.IR.Context)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Context.jl:71
 [30] compile_xla(f::Function, args::Tuple{ConcreteRArray{Float64, 1}}; client::Nothing)
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:573
 [31] compile_xla
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:567 [inlined]
 [32] compile(f::Function, args::Tuple{ConcreteRArray{Float64, 1}}; client::Nothing)
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:600
 [33] compile(f::Function, args::Tuple{ConcreteRArray{Float64, 1}})
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:599
 [34] top-level scope
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:360

julia> Reactant.TracedRArray{Float64, 0}(t::Reactant.TracedRArray{Float64, 0}) = t

julia> ∇f = @compile ∇loss(params′)
┌ Warning: Performing scalar indexing on task Task (runnable) @0x0000744739600010.
│ Invocation resulted in scalar indexing of a TracedRArray.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on device, but very slowly on the CPU,
│ and require expensive copies and synchronization each time and therefore should be avoided.
└ @ Reactant ~/.julia/packages/Reactant/FwbOv/src/TracedRArray.jl:53
┌ Warning: Performing scalar indexing on task Task (runnable) @0x0000744739600010.
│ Invocation resulted in scalar indexing of a TracedRArray.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on device, but very slowly on the CPU,
│ and require expensive copies and synchronization each time and therefore should be avoided.
└ @ Reactant ~/.julia/packages/Reactant/FwbOv/src/TracedRArray.jl:53
┌ Warning: Performing scalar indexing on task Task (runnable) @0x0000744739600010.
│ Invocation resulted in scalar indexing of a TracedRArray.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on device, but very slowly on the CPU,
│ and require expensive copies and synchronization each time and therefore should be avoided.
└ @ Reactant ~/.julia/packages/Reactant/FwbOv/src/TracedRArray.jl:53
┌ Warning: Performing scalar indexing on task Task (runnable) @0x0000744739600010.
│ Invocation resulted in scalar indexing of a TracedRArray.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on device, but very slowly on the CPU,
│ and require expensive copies and synchronization each time and therefore should be avoided.
└ @ Reactant ~/.julia/packages/Reactant/FwbOv/src/TracedRArray.jl:53
┌ Warning: Performing scalar indexing on task Task (runnable) @0x0000744739600010.
│ Invocation resulted in scalar indexing of a TracedRArray.
│ This is typically caused by calling an iterating implementation of a method.
│ Such implementations *do not* execute on device, but very slowly on the CPU,
│ and require expensive copies and synchronization each time and therefore should be avoided.
└ @ Reactant ~/.julia/packages/Reactant/FwbOv/src/TracedRArray.jl:53
ERROR: MethodError: Cannot `convert` an object of type Reactant.TracedRArray{Float64, 0} to an object of type ComplexF64

Closest candidates are:
  convert(::Type{ComplexF64}, ::Basic)
   @ SymEngine ~/.julia/packages/SymEngine/hozAR/src/numerics.jl:187
  convert(::Type{Complex{T}}, ::Basic) where T<:Real
   @ SymEngine ~/.julia/packages/SymEngine/hozAR/src/numerics.jl:191
  convert(::Type{T}, ::ConcreteRArray{T, 0}) where T
   @ Reactant ~/.julia/packages/Reactant/FwbOv/src/ConcreteRArray.jl:73
  ...

Stacktrace:
  [1] setindex!
    @ ./array.jl:1024 [inlined]
  [2] macro expansion
    @ ./multidimensional.jl:960 [inlined]
  [3] macro expansion
    @ ./cartesian.jl:64 [inlined]
  [4] _unsafe_setindex!(::IndexLinear, ::Matrix{ComplexF64}, ::Reactant.TracedRArray{Float64, 0}, ::UnitRange{Int64}, ::UnitRange{Int64})
    @ Base ./multidimensional.jl:955
  [5] _setindex!
    @ ./multidimensional.jl:944 [inlined]
  [6] setindex!
    @ ./abstractarray.jl:1396 [inlined]
  [7] _copy_or_fill!
    @ ./abstractarray.jl:1833 [inlined]
  [8] __cat_offset1!
    @ ./abstractarray.jl:1825 [inlined]
  [9] __cat_offset!(A::Matrix{ComplexF64}, shape::Tuple{Int64, Int64}, catdims::Tuple{Bool, Bool}, offsets::Tuple{Int64, Int64}, x::Reactant.TracedRArray{Float64, 0}, X::Reactant.TracedRArray{Float64, 0})
    @ Base ./abstractarray.jl:1816
 [10] __cat(::Matrix{ComplexF64}, ::Tuple{Int64, Int64}, ::Tuple{Bool, Bool}, ::Reactant.TracedRArray{Float64, 0}, ::Vararg{Reactant.TracedRArray{Float64, 0}})
    @ Base ./abstractarray.jl:1812
 [11] _cat_t
    @ ./abstractarray.jl:1805 [inlined]
 [12] typed_hcat(T::Type, A::Reactant.TracedRArray{Float64, 0}, B::Reactant.TracedRArray{Float64, 0})
    @ Base ./abstractarray.jl:2012
 [13] typed_hvcat(::Type{ComplexF64}, ::Tuple{Int64, Int64}, ::Reactant.TracedRArray{Float64, 0}, ::Vararg{Reactant.TracedRArray{Float64, 0}})
    @ Base ./abstractarray.jl:2175
 [14] mat(::Type{ComplexF64}, R::RotationGate{2, Reactant.TracedRArray{Float64, 0}, YGate})
    @ YaoBlocks ~/.julia/packages/YaoBlocks/DuDXk/src/primitive/rotation_gate.jl:93
 [15] mat(x::RotationGate{2, Reactant.TracedRArray{Float64, 0}, YGate})
    @ YaoBlocks ~/.julia/packages/YaoBlocks/DuDXk/src/abstract_block.jl:120
 [16] Quantum(circuit::ChainBlock{2})
    @ TenetYaoBlocksExt ~/.julia/packages/Tenet/CFIeY/ext/TenetYaoBlocksExt.jl:34
 [17] expectation(params::Reactant.TracedRArray{Float64, 1})
    @ Main ./REPL[11]:9
 [18] loss
    @ ./REPL[12]:2 [inlined]
 [19] (::Tuple{})(none::Reactant.TracedRArray{Float64, 1})
    @ Base.Experimental ./<missing>:0
 [20] (::Reactant.var"#26#35"{typeof(loss), Reactant.MLIR.IR.Block, Vector{Reactant.TracedRArray}, Tuple{Reactant.TracedRArray{Float64, 1}}})()
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/utils.jl:100
 [21] block!(f::Reactant.var"#26#35"{typeof(loss), Reactant.MLIR.IR.Block, Vector{Reactant.TracedRArray}, Tuple{Reactant.TracedRArray{Float64, 1}}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Block.jl:201
 [22] make_mlir_fn(f::Function, args::Tuple{Reactant.TracedRArray{Float64, 1}}, kwargs::Tuple{}, name::String, concretein::Bool; toscalar::Bool)
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/utils.jl:74
 [23] make_mlir_fn
    @ ~/.julia/packages/Reactant/FwbOv/src/utils.jl:24 [inlined]
 [24] autodiff(::ReverseMode{true, false, FFIABI, false, true}, f::EnzymeCore.Const{typeof(loss)}, ::Type{Active}, args::Duplicated{Reactant.TracedRArray{Float64, 1}})
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/Interpreter.jl:253
 [25] autodiff
    @ ~/.julia/packages/Enzyme/8GSlk/src/Enzyme.jl:512 [inlined]
 [26] macro expansion
    @ ~/.julia/packages/Enzyme/8GSlk/src/Enzyme.jl:1705 [inlined]
 [27] gradient
    @ ~/.julia/packages/Enzyme/8GSlk/src/Enzyme.jl:1646 [inlined]
 [28] ∇loss
    @ ./REPL[13]:2 [inlined]
 [29] (::Tuple{})(none::Reactant.TracedRArray{Float64, 1})
    @ Base.Experimental ./<missing>:0
 [30] (::Reactant.var"#26#35"{typeof(∇loss), Reactant.MLIR.IR.Block, Vector{Reactant.TracedRArray}, Tuple{Reactant.TracedRArray{Float64, 1}}})()
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/utils.jl:100
 [31] block!(f::Reactant.var"#26#35"{typeof(∇loss), Reactant.MLIR.IR.Block, Vector{Reactant.TracedRArray}, Tuple{Reactant.TracedRArray{Float64, 1}}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Block.jl:201
 [32] make_mlir_fn(f::Function, args::Tuple{ConcreteRArray{Float64, 1}}, kwargs::Tuple{}, name::String, concretein::Bool; toscalar::Bool)
    @ Reactant ~/.julia/packages/Reactant/FwbOv/src/utils.jl:74
 [33] make_mlir_fn
    @ ~/.julia/packages/Reactant/FwbOv/src/utils.jl:24 [inlined]
 [34] #6
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:253 [inlined]
 [35] block!(f::Reactant.Compiler.var"#6#11"{typeof(∇loss), Tuple{ConcreteRArray{Float64, 1}}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Block.jl:201
 [36] #5
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:252 [inlined]
 [37] mmodule!(f::Reactant.Compiler.var"#5#10"{Reactant.MLIR.IR.Module, typeof(∇loss), Tuple{ConcreteRArray{Float64, 1}}}, blk::Reactant.MLIR.IR.Module)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Module.jl:93
 [38] compile_mlir!(mod::Reactant.MLIR.IR.Module, f::Function, args::Tuple{ConcreteRArray{Float64, 1}}; optimize::Bool)
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:249
 [39] compile_mlir!
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:248 [inlined]
 [40] (::Reactant.Compiler.var"#30#32"{typeof(∇loss), Tuple{ConcreteRArray{Float64, 1}}})()
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:576
 [41] context!(f::Reactant.Compiler.var"#30#32"{typeof(∇loss), Tuple{ConcreteRArray{Float64, 1}}}, ctx::Reactant.MLIR.IR.Context)
    @ Reactant.MLIR.IR ~/.julia/packages/Reactant/FwbOv/src/mlir/IR/Context.jl:71
 [42] compile_xla(f::Function, args::Tuple{ConcreteRArray{Float64, 1}}; client::Nothing)
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:573
 [43] compile_xla
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:567 [inlined]
 [44] compile(f::Function, args::Tuple{ConcreteRArray{Float64, 1}}; client::Nothing)
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:600
 [45] compile(f::Function, args::Tuple{ConcreteRArray{Float64, 1}})
    @ Reactant.Compiler ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:599
 [46] top-level scope
    @ ~/.julia/packages/Reactant/FwbOv/src/Compiler.jl:360
```

@mofeing told me that there are some problems with the 0-dimensinal arrays (scalar), could this be the case?

Thanks!

### PR DESCRIPTION
**Pull request by mofeing**
It's not yet finished, but this should fix #148

### CODE REVIEW

File: src/TracedRArray.jl [hunk 1]
```diff
@@ -759,10 +759,14 @@ function _copyto!(dest::TracedRArray, bc::Broadcasted)
...
+dispatch_val(x) = x
+dispatch_val(::Val{D}) where {D} = D
 
-    # MLIR expects the dimension `D` to be ≤ the rank of the input tensors
+function Base._cat(dims, A::TracedRArray{T,N}, Bs::TracedRArray...) where {T,N}
+    dims = dispatch_val(dims)
+    @assert dims isa Integer "Support for non-integer dimensions is not implemented yet."
+
+    # MLIR expects the dimension `dims` to be ≤ the rank of the input tensors
     A = maybe_expand_dims(A, dims)
```
Comment (avik-pal) [inline src/TracedRArray.jl:770]: `maybe_expand_dims` also needs an update

Comment (mofeing) [reply src/TracedRArray.jl:770]: yep, tests fail on that

File: src/TracedRArray.jl [hunk 2]
```diff
@@ -784,7 +788,8 @@ function Base._cat(dims::Val{D}, A::TracedRArray{T,N}, Bs::TracedRArray...) wher
     return Res
 end
 
-function maybe_expand_dims(x::AbstractArray{T,N}, ::Val{D}) where {T,N,D}
-    D ≤ N && return x
-    return reshape(x, ntuple(i -> i ≤ N ? size(x, i) : 1, Val(D)))
+function maybe_expand_dims(x::AbstractArray, dims) where {T,N}
+    dims = dispatch_val(dims)
+    dims ≤ N && return x
+    return reshape(x, ntuple(i -> i ≤ N ? size(x, i) : 1, dims))
```
Comment (avik-pal) [inline src/TracedRArray.jl:794]:
```suggestion
function maybe_expand_dims(x::AbstractArray{T,N}, dims) where {T,N}
    dims = dispatch_val(dims)
    dims ≤ N && return x
    return reshape(x, ntuple(i -> i ≤ N ? size(x, i) : 1, dims))
```

Comment (mofeing) [reply src/TracedRArray.jl]: we must have committed the same with a difference of seconds jajaja

**Comment by mofeing** [comment]
I was thinking to add some tests, but I'll wait to #161 to land.

**Review by wsmoses — COMMENTED** [review]
Maybe add a test?

**Comment by mofeing** [comment]
> Maybe add a test?

I found that 0-dim arrays have a similar behavior to numbers when using `*cat` methods, so the new tests should be passed by 0-dim arrays even after #161

**Comment by mofeing** [comment]
@avik-pal I don't know why the shapes are wrong. It seems like it's always doing a concatenation of the first direction... but did some print debugging `shape` should be correct.

**Comment by avik-pal** [comment]
```julia
julia> x = fill(true)
0-dimensional Array{Bool, 0}:
1

julia> x_concrete = Reactant.to_rarray(x)
0-dimensional Array{Bool, 0}:
1
```

I don't think it is hitting the dispatches at all

**Comment by mofeing** [comment]
> ```julia
> julia> x = fill(true)
> 0-dimensional Array{Bool, 0}:
> 1
> 
> julia> x_concrete = Reactant.to_rarray(x)
> 0-dimensional Array{Bool, 0}:
> 1
> ```
> 
> I don't think it is hitting the dispatches at all

Fixed! But probably you want to change this `Union` at this line https://github.com/EnzymeAD/Reactant.jl/blob/0520fb77f852987d3deddecb2109366ac8981eda/src/Tracing.jl#L383-L383 for the `PrimitiveTypes` in your PR.

**Comment by avik-pal** [comment]
Yes that PR has the patch

**Comment by mofeing** [comment]
`hcat` and `vcat` are working now; `hvcat` is the only one that fails because it doesn't call `_cat_t`.

**Comment by mofeing** [comment]
@avik-pal it's ready now

**Review by avik-pal — APPROVED** [review]

**Pull request merged**